File: interop/interop.go
package interop

import (
	"context"
	"encoding/json"
	"fmt"
	"math"

	"github.com/complyue/hbi/pkg/he"
)

// ExposeInterOpValues exposes some common names in other languages for interop with them
func ExposeInterOpValues(he *he.HostingEnv) {
	ve := he.AnkoEnv()
	ve.ExecuteContext(context.Background(), `
None = nil
True = true
False = false
null = nil
`)
	ve.Define("NaN", math.NaN())
	ve.Define("nan", math.NaN())
}

// JSONObj converts a map into the json representation as a string
func JSONObj(o map[string]interface{}) string {
	encoded, err := json.Marshal(o)
	if err != nil {
		panic(err)
	}
	return string(encoded)
}

// JSONArray converts a slice into the json representation as a string
func JSONArray(a []interface{}) string {
	encoded, err := json.Marshal(a)
	if err != nil {
		panic(err)
	}
	return string(encoded)
}

// JSONStr converts a string into the json representation as a string
func JSONStr(v interface{}) string {
	s := fmt.Sprintf("%s", v)
	encoded, err := json.Marshal(s)
	if err != nil {
		panic(err)
	}
	return string(encoded)
}

File: interop/interop.py
import json

import hbi

from collections import OrderedDict

__all__ = ["JSONObj", "JSONArray", "JSONStr", "expose_interop_values"]


class JSONObj(OrderedDict):
    """
    A subclass of `collections.OrderedDict`, for JSON object with proper repr and str value in json form.

    """

    def __repr__(self):
        return json.dumps(self)

    __str__ = __repr__


class JSONArray(list):
    """
    A subclass of standard python list, for JSON array with proper repr and str value in json form.

    """

    def __repr__(self):
        return json.dumps(self)

    __str__ = __repr__


class JSONStr(str):
    """
    A subclass of standard python str, for JSON string with proper repr in json form.

    """

    def __repr__(self):
        return json.dumps(self)

    def __str__(self):
        return self


def expose_interop_values(he: "hbi.HostingEnv"):
    globals_ = he.globals

    exec(
        rf"""
from math import nan
NaN = nan

nil = None
true = True
false = False
null = None
""",
        globals_,
        globals_,
    )

    globals_["JSONObj"] = JSONObj
    globals_["JSONArray"] = JSONArray
    globals_["JSONStr"] = JSONStr
